src/cpp/src/gguf_utils/gguf_modeling.cpp

```diff
@@ -0,0 +1,156 @@
+        ov::element::i64, ov::PartialShape{-1, -1});
+    set_name(position_ids, "position_ids");
+
+    auto beam_idx = std::make_shared<ov::op::v0::Parameter>(
+        ov::element::i32, ov::PartialShape{-1});
+    set_name(beam_idx, "beam_idx");
+
+    // Create embedding layer
+    auto [inputs_embeds, embeddings] = make_embedding(
+        "model.embed_tokens",
+        input_ids->output(0),
+        consts,
+        static_cast<QType>(std::get<int>(configs.at("qtype"))));
+
+    auto hidden_states = inputs_embeds;
+
+    // Initialize RoPE
+    auto rope_const = init_rope(
+        std::get<int>(configs.at("head_size")),
+        std::get<int>(configs.at("max_position_embeddings")),
+        std::get<float>(configs.at("rope_freq_base")));
+
+    // Get input shape components
+    auto input_shape = std::make_shared<ov::op::v3::ShapeOf>(input_ids);
+    auto batch_axis = std::make_shared<ov::op::v0::Constant>(
+        ov::element::i64, ov::Shape{1}, 0);
+    auto batch_size = std::make_shared<ov::op::v8::Gather>(
+        input_shape, batch_axis, batch_axis);
+
+    auto hidden_dim = std::make_shared<ov::op::v0::Constant>(
+        ov::element::i64, ov::Shape{1}, 3);
+
+    // Process layers
+    ov::SinkVector sinks;
+    ov::Output<ov::Node> causal_mask;
+    std::pair<ov::Output<ov::Node>, ov::Output<ov::Node>> cos_sin_cached;
+    std::shared_ptr<ov::Node> output_shape = nullptr;
+
+    for (int i = 0; i < std::get<int>(configs.at("layer_num")); ++i) {
+        auto [new_hidden, layer_sinks, new_mask, new_cos_sin, new_shape] = layer(
+            configs,
+            consts,
+            i,
+            hidden_states,
+            attention_mask,
+            causal_mask,
+            position_ids,
+            rope_const,
+            beam_idx,
+            batch_size,
+            hidden_dim,
+            cos_sin_cached,
+            output_shape);
+
+        hidden_states = new_hidden;
+        causal_mask = new_mask;
+        cos_sin_cached = new_cos_sin;
+        output_shape = new_shape;
+
+        sinks.insert(sinks.end(), layer_sinks.begin(), layer_sinks.end());
+    }
+
+    // Final layer norm
+    auto final_norm = make_rms_norm(
+        "model.norm",
+        hidden_states,
+        consts,
+        std::get<float>(configs.at("rms_norm_eps")));
+
+    // LM head
+    auto embed_out = make_lm_head(
+        "lm_head",
+        final_norm,
+        consts,
+        embeddings,
+        static_cast<QType>(std::get<int>(configs.at("qtype"))));
+
+    // Create results
+    auto logits = std::make_shared<ov::op::v0::Result>(embed_out);
+    set_name(logits, "logits");
+
+    // Create model
+    ov::ParameterVector inputs{input_ids, attention_mask, position_ids, beam_idx};
+    auto model = std::make_shared<ov::Model>(ov::OutputVector({logits->output(0)}), sinks, inputs);
+
+    // Set runtime options
+    model->set_rt_info("f16", {"runtime_options", "KV_CACHE_PRECISION"});
```

Comment:
do we need fp16 precision for quantized models? should it be a condition here?

Reply:
Fixed